### PR TITLE
Add multi value processing to GetQueryString

### PR DESCRIPTION
**Pull request by cadatoiva**
This adds support for using NameValueCollection with query strings that have multiple values but doesn't support NameValueCollection's way of storing those values.

When you add a second value to an existing key, like in #7913, it gets stored in NVC as `pairs["cat[]"] == "19,6";` This change allows us to do `pairs.GetQueryString(splitMultiValues: true)` to get the desired effect. It's not just that PR, I was already implementing this for a change to NCore I'm trying to do so that we can still use `pairs["key"]` instead of trying to remember the index in a `List<KeyValuePair>` to change something in multiple requests (multi-page processing as example).

I also made it so we can choose the separator between values, so we can use this for post requests, where the query is separated by `;` instead of `&`

The default values for both of these new params leaves all existing uses working as intended.

### CODE REVIEW

File: src/Jackett.Common/Utils/StringUtil.cs [hunk 1]
```diff
@@ -138,13 +138,35 @@ public static string MakeValidFileName(string text, char? replacement = '_', boo
...
-            string.Join("&", collection.AllKeys.Select(a =>
-                $"{a}={WebUtilityHelpers.UrlEncode(collection[a], encoding ?? Encoding.UTF8)}"));
+        /// <summary>
+        /// Converts a NameValueCollection to an appropriately formatted query string.
+        /// Duplicate keys are allowed in a NameValueCollection, but are stored as a csv string in Value.
+        /// This function handles leaving the values together in the csv string or splitting the value into separate keys
+        /// </summary>
+        /// <param name="collection">The NameValueCollection being converted</param>
+        /// <param name="encoding">The Encoding to use in url encoding Value</param>
+        /// <param name="splitMultiValues">Whether the Value should be Key=Value or split into Key=Val1&Key=Val2</param>
```
Comment (ngosang) [inline src/Jackett.Common/Utils/StringUtil.cs]: Whether the Value should be Key=Val1,Val2 or split into Key=Val1&Key=Val2

File: src/Jackett.Common/Utils/StringUtil.cs [hunk 2]
```diff
@@ -138,13 +138,35 @@ public static string MakeValidFileName(string text, char? replacement = '_', boo
...
+        /// <param name="separator">The string used to separate each query value</param>
+        /// <returns>A web encoded string of key=value parameters separated by the separator</returns>
+        public static string GetQueryString(this NameValueCollection collection, Encoding encoding = null,
```
Comment (ngosang) [inline src/Jackett.Common/Utils/StringUtil.cs]: Encoding encoding = Encoding.UTF8
and remove `?? Encoding.UTF8` below

File: src/Jackett.Common/Utils/StringUtil.cs [hunk 3]
```diff
@@ -138,13 +138,35 @@ public static string MakeValidFileName(string text, char? replacement = '_', boo
...
+                separator,
+                collection.AllKeys.Select(
+                    a => $"{a}={WebUtilityHelpers.UrlEncode(collection[a], encoding ?? Encoding.UTF8)}"));
+        }
 
-        public static string GetQueryString(this ICollection<KeyValuePair<string, string>> collection, Encoding encoding = null) =>
-            string.Join("&", collection.Select(a =>
-                $"{a.Key}={WebUtilityHelpers.UrlEncode(a.Value, encoding ?? Encoding.UTF8)}"));
+        public static string GetQueryString(this IEnumerable<KeyValuePair<string, string>> collection,
+                                            Encoding encoding = null, string separator = "&") =>
```
Comment (ngosang) [inline src/Jackett.Common/Utils/StringUtil.cs]: same about encoding

File: src/Jackett.Common/Utils/StringUtil.cs (hunk 2 above)
Comment (ngosang) [inline src/Jackett.Common/Utils/StringUtil.cs]: It's possible to extend NameValueCollection instead?

File: src/Jackett.Common/Utils/StringUtil.cs [hunk 4]
```diff
@@ -138,13 +138,35 @@ public static string MakeValidFileName(string text, char? replacement = '_', boo
             return changed ? sb.ToString() : text;
         }
 
-        public static string GetQueryString(this NameValueCollection collection, Encoding encoding = null) =>
-            string.Join("&", collection.AllKeys.Select(a =>
-                $"{a}={WebUtilityHelpers.UrlEncode(collection[a], encoding ?? Encoding.UTF8)}"));
+        /// <summary>
```
Comment (ngosang) [inline src/Jackett.Common/Utils/StringUtil.cs]: please, add some unit tests

File: src/Jackett.Common/Utils/StringUtil.cs (hunk 2 above)
Comment (cadatoiva) [reply src/Jackett.Common/Utils/StringUtil.cs]: We should either keep `?? Encoding.UTF8` or change it to `?? throw new ArgumentNullException(nameof(encoding) + " cannot be null")` in case of someone passing in null explicitly. But we can change the default value so it shows up in the tooltip for the function.

Comment (cadatoiva) [reply src/Jackett.Common/Utils/StringUtil.cs]: Not sure what you're asking? NameValueCollection is a Microsoft built-in class in System.Collections.Specialized.

Comment (ngosang) [reply src/Jackett.Common/Utils/StringUtil.cs]: I guess `WebUtilityHelpers.UrlEncode(` will throw the exception for us if it's not valid encoding or null

Comment (cadatoiva) [reply src/Jackett.Common/Utils/StringUtil.cs]: Actually, this is null because we can't use `Encoding.UTF8` as a default value. Compilation error on the change says `Default parameter must be a compile time constant` which `Encoding.UTF8` is not.

Comment (ngosang) [reply src/Jackett.Common/Utils/StringUtil.cs]: I explained bad myself. How this will be called?
`var qc = new NameValueCollection ...    qc.GetQueryString()`
or
`var qc = new NameValueCollection ...    StringUtil.GetQueryString(qc)`
I don't like the 2nd one, it's error prone.

Comment (cadatoiva) [reply src/Jackett.Common/Utils/StringUtil.cs]: This is the extension method we've been using in this pattern:
```
var queryParams = new NameValueCollection
{ ... };
var searchLink = SearchLink + "?" + queryParams.GetQueryString();
```

So to answer your question, it is called in the first manner. Although technically you can call it the second way too, because the compiler is converting the first to the second behind the scenes. Any static function in a static class whose first parameter is `this var name` can be used like this.

Comment (ngosang) [reply src/Jackett.Common/Utils/StringUtil.cs]: oki

Comment (ngosang) [reply src/Jackett.Common/Utils/StringUtil.cs]: ohh, I like this :)

File: src/Jackett.Common/Utils/StringUtil.cs [hunk 5]
```diff
@@ -138,13 +138,35 @@ public static string MakeValidFileName(string text, char? replacement = '_', boo
...
+        /// <returns>A web encoded string of key=value parameters separated by the separator</returns>
+        public static string GetQueryString(this NameValueCollection collection, Encoding encoding = null,
+                                            bool splitMultiValues = false, string separator = "&")
+        {
+            if (splitMultiValues)
+                return collection.AllKeys.SelectMany(
+                                     key => collection[key]
+                                            .Split(',')
+                                            .Select(value => new KeyValuePair<string, string>(key, value)))
+                                 .GetQueryString(encoding);
```
Comment (ngosang) [inline src/Jackett.Common/Utils/StringUtil.cs]: you miss `, separator` ?

File: src/Jackett.Common/Utils/StringUtil.cs [hunk 6]
```diff
@@ -138,13 +138,35 @@ public static string MakeValidFileName(string text, char? replacement = '_', boo
...
+        /// </summary>
+        /// <param name="collection">The NameValueCollection being converted</param>
+        /// <param name="encoding">The Encoding to use in url encoding Value</param>
+        /// <param name="splitMultiValues">Whether the Value should be Key=Value or split into Key=Val1&Key=Val2</param>
+        /// <param name="separator">The string used to separate each query value</param>
+        /// <returns>A web encoded string of key=value parameters separated by the separator</returns>
+        public static string GetQueryString(this NameValueCollection collection, Encoding encoding = null,
+                                            bool splitMultiValues = false, string separator = "&")
+        {
+            if (splitMultiValues)
```
Comment (ngosang) [inline src/Jackett.Common/Utils/StringUtil.cs]: this var name is not clear to me. what about `commaSeparatedMultiValues` instead?

File: src/Jackett.Common/Utils/StringUtil.cs [hunk 7]
```diff
@@ -138,13 +138,35 @@ public static string MakeValidFileName(string text, char? replacement = '_', boo
...
+        /// Duplicate keys are allowed in a NameValueCollection, but are stored as a csv string in Value.
+        /// This function handles leaving the values together in the csv string or splitting the value into separate keys
+        /// </summary>
+        /// <param name="collection">The NameValueCollection being converted</param>
+        /// <param name="encoding">The Encoding to use in url encoding Value</param>
+        /// <param name="splitMultiValues">Whether the Value should be Key=Value or split into Key=Val1&Key=Val2</param>
+        /// <param name="separator">The string used to separate each query value</param>
+        /// <returns>A web encoded string of key=value parameters separated by the separator</returns>
+        public static string GetQueryString(this NameValueCollection collection, Encoding encoding = null,
+                                            bool splitMultiValues = false, string separator = "&")
```
Comment (ngosang) [inline src/Jackett.Common/Utils/StringUtil.cs]: We are changing the default behavior, this is ok but I think you should take a look at the current usages..

Comment (cadatoiva) [reply src/Jackett.Common/Utils/StringUtil.cs]: The default behavior remains the same as it was. With `splitMultiValues == false` and `separator == "&"` the code that runs is:
```
return string.Join(separator/*=="&"*/, collection.AllKeys.Select(
                    a => $"{a}={WebUtilityHelpers.UrlEncode(collection[a], encoding ?? Encoding.UTF8)}"));
```
which is the original setup of `Key=Val1,Val2` and the reason why you couldn't use NameValueCollection in your PR.

File: src/Jackett.Common/Utils/StringUtil.cs (hunk 6 above)
Comment (cadatoiva) [reply src/Jackett.Common/Utils/StringUtil.cs]: `duplicateKeyIfMultipleValues`?

**Comment by cadatoiva** [comment]
I needed to move the key split code into a separate function for use in NCore, otherwise the whole point of this improvement is lost for my refactoring. Unit tests have been added that cover all code paths.

File: src/Jackett.Test/Utils/StringUtilTests.cs [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+                { "st", "troy" },
+                { "makin", "thisup" },
+                { "st", "duplicate" }
+            };
+            const string combinedTemplate = "q=test{0}st=troy%2Cduplicate{0}makin=thisup";
+            const string splitTemplate = "q=test{0}st=troy{0}st=duplicate{0}makin=thisup";
+            Assert.AreEqual(string.Format(combinedTemplate, "&"), testCase.GetQueryString());
+            Assert.AreEqual(string.Format(combinedTemplate, ";"), testCase.GetQueryString(separator: ";"));
+            Assert.AreEqual(string.Format(splitTemplate, "&"), testCase.GetQueryString(duplicateKeysIfMulti: true));
+            Assert.AreEqual(
```
Comment (ngosang) [inline src/Jackett.Test/Utils/StringUtilTests.cs]: Please add more tests. Encoding and word with spaces are missing.
* q = españa (encoding utf-8)
* q = Ру́сский (utf-8)
* q = Ру́сский (windows-1251)
* q = Harry Potter (utf-8)  // word with spaces
* q = dark&night (utf-8)  // word contains separator

Also add some "bad tests". We are not failing in some cases in the code...
* Empty and null NameValueCollection
* Empty and Null separator
* Null encoding

No need to skimp, the test code is not part of the final binary. Nor does the code need to be highly optimized, it is more important that it is easy to understand and maintain. Create more NameValueCollection vars, if you reuse the same all the time it's really hard to add more tests. Also add comments to explain what you are testing.

File: src/Jackett.Common/Utils/StringUtil.cs [hunk 9]
```diff
@@ -138,16 +138,41 @@ public static string MakeValidFileName(string text, char? replacement = '_', boo
...
-            string.Join("&", collection.Select(a =>
-                $"{a.Key}={WebUtilityHelpers.UrlEncode(a.Value, encoding ?? Encoding.UTF8)}"));
+        /// <summary>
+        /// Converts a NameValueCollection to an appropriately formatted query string.
+        /// Duplicate keys are allowed in a NameValueCollection, but are stored as a csv string in Value.
+        /// This function handles leaving the values together in the csv string or splitting the value into separate keys
+        /// </summary>
+        /// <param name="collection">The NameValueCollection being converted</param>
+        /// <param name="encoding">The Encoding to use in url encoding Value</param>
+        /// <param name="splitMultiValues">Duplicate keys are handled as true => {"Key=Val1", "Key=Val2} or false => {"Key=Val1,Val2"}</param>
```
Comment (ngosang) [inline src/Jackett.Common/Utils/StringUtil.cs]: duplicateKeysIfMulti

File: src/Jackett.Test/Utils/StringUtilTests.cs [hunk 10]
```diff
@@ -0,0 +1,28 @@
+using System.Collections.Specialized;
+using NUnit.Framework;
+
+namespace Jackett.Common.Utils.Tests
+{
+    [TestFixture]
+    public class StringUtilTests
+    {
+        [Test]
```
Comment (ngosang) [inline src/Jackett.Test/Utils/StringUtilTests.cs]: Add another test for `ToEnumerable` method

File: src/Jackett.Common/Utils/StringUtil.cs (hunk 6 above)
Comment (ngosang) [reply src/Jackett.Common/Utils/StringUtil.cs]: ok

**Comment by ngosang** [comment]
@cadatoiva you abandon this? 😢

**Comment by cadatoiva** [comment]
I didn't abandon it, I had switched branches to work on TV Store and it turned out being much bigger than I thought it was going to be, and I couldn't commit/stash and pause in the middle like I could with others. I'm coming back to this next, I just have to figure out how to structure the tests you've requested.

**Comment by cadatoiva** [comment]
Ok, I've made the suggested changes above. I took inspiration from https://github.com/Jackett/Jackett/blob/master/src/Jackett.Test/WebUtilityHelpersTests.cs to generate the test cases across multiple encodings and q values.

File: src/Jackett.Test/Utils/StringUtilTests.cs (hunk 8 above)
Comment (ngosang) [reply src/Jackett.Test/Utils/StringUtilTests.cs]: @cadatoiva Please, don't take this as an attack, I don't want to argue, if you think it's okay you can merge.

I see the changes and I have no doubts you are putting effort in this, but that test is useless to me. I spent 15 minutes looking at the test and I'm still not sure if you are testing everything. The test is much more hard to understand that the code itself. This is not how test are done.

You have to see the test as reference documentation with simple use case, just input and output, the tested function is a blackbox. Please, don't try to reuse code. That makes the code hard to maintain if you add more functionality in the tested function. Also don't try to generate the expected response using string. Format or something like that, just write the string by hand. You don't have to cover all permutations of all parameters. I expect the test to look like this:
```
var testCase = new NameValueCollection
{
    { "q", "test" },
    { "st", "troy" },
    { "makin", "thisup" },
    { "st", "duplicate" }
};

// base test
Assert.AreEqual("q=test&st=troy%2Cduplicate&makin=thisup", testCase.GetQueryString());

// testing separator
Assert.AreEqual("q=test;st=troy%2Cduplicate;makin=thisup", testCase.GetQueryString(separator: ";"));

// testing duplicateKeysIfMulti
Assert.AreEqual("q=test&st=troy&st=duplicate&makin=thisup", testCase.GetQueryString(duplicateKeysIfMulti: true));

// testing encoding

// testing words with spaces

// testing words with separator

// testing all parameters together

// testing invalid parameters
 => Empty and null NameValueCollection
 => Empty and Null separator
 => Null encoding
```

Comment (cadatoiva) [reply src/Jackett.Test/Utils/StringUtilTests.cs]: I'm not trying to argue back, I just want to make sure we're on the same page here with this. I mentioned in a different comment that I used another test in our project to model this one off of for your additional use cases. Specifically, https://github.com/Jackett/Jackett/blob/master/src/Jackett.Test/WebUtilityHelpersTests.cs

I personally feel like some of these additional test cases are violating the black box principle of how tests are supposed to be used. This particular function is designed to create a web encoded string from a NameValueCollection. We already have a test to make sure that the web encoding part is handled correctly (linked above) so testing spaces and separators in the values feels redundant. The code as is doesn't check to make sure the separator isn't used inside the values, or gets removed from the component strings.

Especially if our test is supposed to be a pseudo-documentation for people to look at, using separators like `{"&", ";", " ", string.Empty, null}` and checking that they are converted when put inside value strings could lead some new developers to think that this code checks and converts separators wholesale, instead of as a side-effect of the web encoding of strings. What happens when someone uses `GetQueryString(separator: "f")`? They might expect from our test cases that "f" would be converted to a "%xx" if found outside of separator usage, when really all we're doing in the test is basically doubling up on testing `WebUtilityHelpers.UrlEncode(string)`

I have an idea on what to do to make it more readable and extensible if we add more features, but I'm probably going to reduce the test cases to better test to just the following:

* Encoding is applied appropriately - 3 tests with a string that has a different URL safe encoding for the two encodings used
  * Default encoding to make sure the default remains UTF-8 to avoid breaking default dependant code.
  * windows-1251 to make sure the non-default encoding is used
  * that an explicit null encoding properly defaults to UTF-8, as is the current setup in case the way we specify the default value changes.
* Separator is applied appropriately - 2 separators 
  * default to make sure the current default of "&" is used to avoid breaking default dependant code
  * ";" to make sure the current default is correctly overridden
  * I feel any other values like `null` or `string.Empty` are just testing that `string.Join` is doing its job, and outside the test scope. The current implementation does not error on null, empty or whitespace separators (per the implementation of `string.Join`) nor do I think it should.
* Keys are split when asked - 3 tests
  * default to make sure keys are not split on the default value so we don't break current base depending on default values
  * `true` to test keys are correctly split
  * `false` to make sure keys are correctly not split
* Testing NameValueCollection specifically - 2 tests
  * an empty NameValueCollection in the current implementation returns `string.Empty` and should remain that way in any future implementations. This was true for the original code as well.
  * a null NameValueCollection throws an exception. 

Some kind of combinatorial testing (what you called "all parameters at once"?) should possibly also be tested to make sure that all the above tests work on all code paths (e.g. the Encoding doesn't get applied when splitting keys, but does when they aren't). I'm not sure on this one, though, because if we go on to extend functionality, then it doesn't guarantee that our current combinatorial tests will cover all code paths. Plus I think I read somewhere that this is outside the scope of Unit Testing and is for Integration Testing instead.

I welcome any feedback or corrections you have on what I said, as I only have a theory level knowledge of Unit Testing best practices, and have very little experience in actually using them.

Comment (ngosang) [reply src/Jackett.Test/Utils/StringUtilTests.cs]: > The code as is doesn't check to make sure the separator isn't used inside the values, or gets removed from the component strings.

> I feel any other values like null or string.Empty are just testing that string.Join is doing its job, and outside the test scope. The current implementation does not error on null, empty or whitespace separators (per the implementation of string.Join) nor do I think it should.

We are not in the same page here.
1. The goal of this function is to create a VALID querystring. If the user is setting some parameters that will generate INVALID querysting (according to HTTP RFC) we must fail. That querystring will be part of one URL and if it's not valid something will fail later. It's better to detect the problems ASAP.
2. You are not seeing the function as a blackbox because you are assuming the function is using Join and Join does... What happens if I change the Join with other function in the future? Are your tests still valid? I don't think so, because that function will do different things with empty/null. Just think as inputs and outputs, as you don't know the internals. That's is why TTD is so popular, you write the test before the function so the tests are not biased by your knowledge of the function.
3. After years of coding my workflow is: I write the tests first (or at least I write the real code with tests in mind). In the test I don't include all the infinite combination of inputs, just the cases I think will be problematic. For example, empty/null strings sometimes causes NullPointerException. So in the test I test the bad paths too, in this case I expect if the separator is null, and exception to be thrown. When I write the real code, the test will remind me that I forget to throw the exception in that case. In conclusion, think about desired functionality, write or think in the test and change the real code until it pass the tests, TDD. **In this case add the tests for null/empty/whitespace separator and throw the exception.**

> Some kind of combinatorial testing (what you called "all parameters at once"?) should possibly also be tested to make sure that all the above tests work on all code paths (e.g. the Encoding doesn't get applied when splitting keys, but does when they aren't).

This is necessary only if we have different code paths in our code. I don't think this is necessary for this function. **Code coverage tools** will tell you if there are code paths not covered by tests, I don't know Visual Studio but I'm pretty sure it has a tool for that. 
The difference between Unit/**Integration test** is if you have external dependencies. In this case is a Unit test because you are testing the function isolated. If this function makes a call to a remote URL or the if frontend has to be tested in a real web browser (with Selenium) these are Integration tests. You need a external tools/systems to make the test.

I agree with the test you are proposing (just add the bad separator cases).
My major concern is you are obsessed with perfect code and you are writing unredable test code, the purpose of the tests are not performance is to catch bugs, easy maintainability and be pseudo-documentation. It's allowed duplicate code, write large comments, large methods... most projects have style checkers and code coverage disabled for tests for that reason. Please, follow the example I let before.

Comment (cadatoiva) [reply src/Jackett.Test/Utils/StringUtilTests.cs:70]: > 1. The goal of this function is to create a VALID querystring. If the user is setting some parameters that will generate INVALID querysting (according to HTTP RFC) we must fail. That querystring will be part of one URL and if it's not valid something will fail later. It's better to detect the problems ASAP.

The changes I've made have been put in place to be able to use this function outside the confines of a URI querystring, and into any place where a URL encoded `key=value` string can be used in web requests. That's why you see the test cases including the use of the separator `;` as the alternate test unit, so it can be used for post data or cookies (along with the separator `,` for some cookies, but that's outside my immediate need). It is possible that there are additional use cases that neither of us are thinking of (SQL, JS, JSON, AJAX, or standards changing, human readable formats for error logs, etc) which is why I purposefully left it versatile, instead of throwing errors outside of our imagined use cases.

For point 2, I could see testing to make sure that empty/null creates a zero width delimiter in the string in case we do move away from string.Join (why would we though?) to ensure the replacement behaves the same way as string.Join in these rarely requested cases. BUT I can also see just leaving this use case as an undefined behavior, and not testing for this rare edge case at all. The current implementation would require someone to explicitly ask for null/empty delimiter (which I can only assume that the coder using it for that purpose has good reason to, even if I can't imagine why), and leaving it out of the test would signal that you cannot expect the behavior of the function in an empty delimited string list to remain constant. It doesn't throw an error, because someone might want it (why though?) but it also is subject to change at any point if for some reason we move off of string.Join for delimitation (again, why?). If the null/empty edge case ever does find a legitimate use within our base, whoever does so could add tests to make sure future changes to the function don't change their expected result.

As to point 3 in my years of coding, this is the only project I've written any kind of Unit Testing for. And my research into the topic of Unit Testing has included not just the kinds of articles that say how and why it's good, but also temper the idea with examples of where it shouldn't be used or necessarily required for every function we make such as the following below. These articles also offer guidelines that suggest that these literal one line lambda functions are probably on the list of "really shouldn't require unit tests," but I've personally been behind the keyboard of "simple" one liners being refactored and inadvertently changing functionality that might have been caught by some unit tests, so I haven't been pushing back on that front.

- https://medium.com/pacroy/why-most-unit-testing-is-waste-tests-dont-improve-quality-developers-do-47a8584f79ab
- https://dzone.com/articles/unit-testing-guidelines-what-to-test-and-what-not
- https://techbeacon.com/app-dev-testing/no-1-unit-testing-best-practice-stop-doing-it
- https://www.javaworld.com/article/2892225/how-just-about-everyone-gets-unit-testing-wrong.html

I guess what I'm trying to do while writing these tests isn't unit testing per se, but is instead a Unit level regression test. I want to make sure the code we have in the ways we are actively (or soon to be) using it doesn't change while we refactor the core code base. I want to make sure we aren't breaking things while we stumble through this code base trying to clean it up, or when we add new features, or move things around or combine classes.

Comment (ngosang) [reply src/Jackett.Test/Utils/StringUtilTests.cs:70]: > BUT I can also see just leaving this use case as an undefined behavior, and not testing for this rare edge case at all. 

This is where I wanted to go, don't focus on Join, no one is going to change that code. If the function is a black box, it has to have a defined behavior for all the inputs. I don't care much if we return a malformed querystring or an exception, but we have to return the same in the future. Imagine that someone is using this function with empty separator because it fits his needs. Months later I change the method to add more parameters or something and I change the behavior when the separator is empty. The tests will pass and I will say with confidence my code is working, but I broke the other user code.
If you do test you have to do it well, because if not is worse than when you don't have tests. If I don't have tests I will be more meticulous when testing the new code. Just test all cases, it's only 2 lines of code...

> As to point 3 in my years of coding, this is the only project I've written any kind of Unit Testing for

Bad practice. In some business like consultancy, there are no tests because they are not thinking in the long term maintainability, just in short term profits. I work in a huge multinational and test are mandatory, in the space sector the standard is each line of code has to have unit test + integration tests + acceptance tests (all automated). I also collaborate in the Linux kernel, LibreOffice, Wikimedia... and tests are mandatory. With a large code base, several developers and no test I guarantee you that you spend more time fixing bugs than adding new features.

> I guess what I'm trying to do while writing these tests isn't unit testing per se, but is instead a Unit level regression test.

Bingo. When you are writing the tests at the same time you are writing the code, you know well the function behavior and you are "defining/setting" it in the tests. If someone changes the behavior in the future the test will fail and he will have to understand the test and fix the code or update the test.
Writing tests will also make you a better coder in several ways. By design, if your code is well written, it is easy to test if not you are fucked up. Is usual to discover bugs when you are writing tests. 

I want to improve the test coverage for the core, but it will take time. Just trust me in this, it's necessary.

To merge this PR, you don't have to change the behavior, just add 3 more tests (they are free). Assert the current return value.
* key and value with spaces
* empty separator
* null separator

Comment (cadatoiva) [reply src/Jackett.Test/Utils/StringUtilTests.cs:79]: > key and value with spaces
Current functionality doesn't make keys web-safe, only values. I assume this is on purpose to avoid changing the key in `queryParams["cat[]"] = category;` as a specific example. As none of our current use cases require escaping the keys, I will add this test without changing the behavior as you've requested. Just want you to be aware of this.

> Bad practice
While I do agree that exactly 0 unit tests is probably a bad practice, we're going to have to agree to disagree on the whether 100% coverage by Unit Tests is a good practice. There are a significant number of authoritative sources on both sides of that discussion, some of which I linked to above.
